text/0000-structural-records.md

```diff
@@ -0,0 +1,1170 @@
+structural records. It is perfectly valid to move or copy:
+
+```rust
+println!("The answer to life... is: {}", my_record.alpha);
+```
+
+or to borrow a field:
+
+```rust
+fn borrow(x: &bool) { .. }
+
+borrow(&my_record.beta);
+```
+
+including mutably:
+
+```rust
+fn mutably_borrow(x: &mut bool) { .. }
+
+let mut my_record = { alpha: 42u8, beta: true, gamma: "Dancing Ferris" };
+
+mutably_borrow(&mut my_record.beta);
+```
+
+### One-field records
+
+Positional tuples with a single element use the syntax `(x,)` both for types
+and for values to distinguish between the 1-element tuple and wrapping a value
+in parenthesis. To avoid ambiguities with the `{ x }` syntax that is already
+allowed, we impose the same rule for one-field structural records. To create
+such a record, you must write `{ x, }` or `{ x: <value>, }`. This difference
+is also imposed to maintain a sense of consistency between the positional
+and the non-positional structural variant.
+
+### Struct update syntax
+
+Nominal structs support what is referred to as the *"struct update syntax",
+otherwise known as functional record update (FRU). For example, you can write:
+
+```rust
+struct Color {
+    red: u8,
+    green: u8,
+    blue: u8,
+}
+
+let yellow = Color { red: 0, green: 255, blue: 255 };
+
+let white = Color { red: 255, ..yellow };
+```
+
+This also works for structural records, so you can write:
+
+```rust
+let yellow = { red: 0, green: 255, blue: 255 };
+
+let white = { red: 255, ..yellow };
+```
+
+To match the behaviour of FRU for nominal structs, we impose a restriction
+that the fields mentioned before `..yellow` must all exist in `yellow` and
+have the same types. This means that we cannot write:
+
+```rust
+let white_50_opacity = { alpha: 0.5, red: 255, ..yellow };
+```
+
+However, there is no fundamental reason why we could not allow this,
+but to start off more conservative and to be uniform in behaviour,
+we don't allow this for the time being.
+
+### The type of a structural record
+
+You might be wondering what the type of `my_record` that we've been using
+thus far is. Because this is structural typing, the fields are significant,
+so the type of the record is simply:
+
+```rust
+type TheRecord = { alpha: u8, beta: bool, gamma: &'static str };
+
+let my_record: TheRecord = { alpha: 42u8, beta: true, gamma: "Dancing Ferris" };
+```
+
+Notice how this matches the way we defined `MyRecordType` is we remove
+the prefix `struct MyRecordType`.
+
+The order in which we've put `alpha`, `beta`, and `gamma` here does not matter.
+We could have also written:
+
+```rust
+type TheRecord = { beta: bool, alpha: u8, gamma: &'static str };
+```
+
+As long as the type is a *permutation* of each pair of field name and field type,
+the type is the same. This also means that we can write: 
+
+```rust
+let my_record: TheRecord = { alpha: 42u8, gamma: "Dancing Ferris", beta: true };
+```
+
+### Implemented traits
```

Comment:
Is there any way to apply custom traits for these types e.g. `impl serde::Serialize for {x: u16, y: u32}`

Reply:
OK... Perhaps add this to the Drawbacks or Future possibilities section.